msal_extensions/_cache_lock.py

```diff
@@ -0,0 +1,41 @@
+import os
+import errno
+import datetime
+import psutil
+import portalocker
+
+
+class CrossPlatLock(object):
+    """
+    Offers a mechanism for waiting until another process is finished interacting with a shared resource. This is
+    specifically written to interact with a class of the same name in the .NET extensions library.
+    """
+    TIMEOUT = datetime.timedelta(minutes=1)
+    RETRY_WAIT = datetime.timedelta(milliseconds=100)
+    RETRY_COUNT = int(TIMEOUT.total_seconds() / RETRY_WAIT.total_seconds())
+
+    def __init__(self, lockfile_path):
+        self._lockpath = lockfile_path
+
+    def __enter__(self):
+        pid = os.getpid()
+        proc = psutil.Process(pid)
+        lock_dir = os.path.dirname(self._lockpath)
+        if not os.path.exists(lock_dir):
+            os.makedirs(lock_dir)
+
+        self._fh = open(self._lockpath, 'wb+', buffering=0)
+        portalocker.lock(self._fh, portalocker.LOCK_EX)
+        self._fh.write('{} {}'.format(pid, proc.name()).encode('utf-8'))
+
+    def __exit__(self, *args):
+        self._fh.close()
+        try:
+            # Attempt to delete the lockfile. In either of the failure cases enumerated below, it is likely that
```

Comment:
It looks good, but from a feature requirement's standpoint, is such removal really necessary? It seems we can leave it there. The .Net extension seems to also [leave it there](https://github.com/AzureAD/microsoft-authentication-extensions-for-dotnet/blame/master/src/Shared/CrossPlatLock.cs#L76-L80).

Reply:
Aha, this one I do feel good about! .NET is just sneaky and signals that the file should be deleted when the file handle is first constructed here:
https://github.com/AzureAD/microsoft-authentication-extensions-for-dotnet/blob/113ecc725891395dd7212aac82d0681f9a0d60ce/src/Shared/CrossPlatLock.cs#L53

For that reason, I believe this should stay as is.

Reply:
We had a in-depth offline conversation.

* We believe the motivation of having file removal behavior was partially driven by the convention of writing some debug message into the lock file, which is in-itself technically unnecessary for pure locking purpose.
* And we are aware that the extra file removal behavior here creates chances for racing conditions between file unlock and file removal. And then it forces us to catch and skip those exceptions in this implementation. It won't be necessary if we just use raw locking behavior.

We may revisit this in future.

Reply:
We chatted about this a bit offline, and came to the conclusion that there are some benefits to not deleting this file. However, we left deciding that it's an issue we should take up with the .NET behavior and have a continued conversation about it, instead of blocking the PR on it.